### DISCUSSION TITLE
Defining a homogeneous TypedDict

### DISCUSSION THREAD

**Discussion by maflAT** [python/typing#1592] [Q&A] [answered]
One of my favorite new constructs since 3.12 is a generic, but homogenious TypedDict like this:
```python
class StructuredMap[T](TypedDict):
    a: T
    b: T
``` 

My problem is that, in contrast to a regular dict, this construct does not propagate its type specialization to `values()` and `items()`.
`values()` and `items()` will always be inferred as  `dict_values[str, object]` and `dict_items[str, object]` respectively.

In other words, what i want to define is a regular dict (with homogeneous value types), but with pre-defined keys.

Is there a way to annotate such a construct? Am I thinking too complicated?

**Comment by erictraut** [comment, marked as answer]
There's currently no way to define that in the type system. There is a draft [PEP 728](https://github.com/python/peps/pull/3441) that would allow for this. The reason that a type checker cannot narrow the type of `values` and `items` with today's `TypedDict` is because other (unspecified) keys may be present in a class that is compatible with a `TypedDict`. Since these potential other keys have values of an unknown type, they need to be treated as `object`. PEP 728 would allow you to specify that no other keys can be present.

**Reply by Akuli** [reply]
Maybe type checkers could support this today with a `@final`-decorated TypedDict.

```python
from typing import final, TypedDict

@final
class Foo(TypedDict):
    x: str
    y: str

foo: Foo = {"x": "blah", "y": "blah"}
key, value = list(foo.items())[0]
reveal_type(value)  # object, but maybe could be str instead?
```

With total TypedDicts (the default), there cannot be extra fields at all, so there cannot be extra fields with the wrong type.

**Reply by erictraut** [reply]
`TypedDict` is a structural type, and `@final` is meaningless for structural types. It simply means "you cannot subclass from this class".

> With total TypedDicts (the default), there cannot be extra fields at all

That's incorrect. Totality simply means that all of the defined fields are required, not optional. It doesn't mean that additional fields cannot be present.
